### ISSUE TITLE
C-n suggestions replace same word in the file at somewhere *sometimes*

### ISSUE THREAD

**Issue by xnuk** [kak-lsp/kak-lsp#641]
https://user-images.githubusercontent.com/3071003/187028102-498aec1e-46f3-4faa-b50a-16031b84e583.mp4

Kakoune v2021.11.08, kak-lsp 12.1.0-snapshot, Rust rls 1.41.0, on Alacritty 0.10.1, on Sway 1.7, on Arch linux x64.

Config files: https://github.com/xnuk/--/tree/b1763cf417be2ba56fd1e8ef7ebb72a034d24ab8/kak

This happens when:
- The code is somehow complex enough; years of kak-lsp usage experiencing same problem and still couldn't find minimal reproducible steps
- Selection exists before entering insert mode, like pressing `a`. In other words: `%val{selections_length}` is not exactly `1`.

**Comment by krobelus** [comment]
If it happens, does it happen consistently? That is, can you undo and trigger the problem again, or is it random?
A non-minimal reproducer is fine as well. If possible, commit and push the code that triggers the problem so I can try to reproduce.

Really interesting bug.

**Comment by xnuk** [comment]
Okay I think I made a reproducible one, but unsure if it happens in other computer: https://github.com/xnuk/kakoune-likes-to-replace-word Hope this reproduce well.

> If it happens, does it happen consistently? That is, can you undo and trigger the problem again, or is it random?

With same code and same configuration, yes. Same thing happens again and again, at that specific position. In my experience, same thing happens with `tsserver`-like LSP server configuration.

**Comment by krobelus** [comment]
wow nice job. Unfortunately it doesn't happen here even though I used exactly the same software versions as you.

Can you post a debug log? That might tell us more

**Comment by xnuk** [comment]
Ok, here's the log: https://gist.github.com/xnuk/fec6a66a82c46746d8158756b81b2ab0

Got from:
```kak
set global lsp_cmd "kak-lsp -s %val{session} -vvvv --log /tmp/kak-lsp.log --config=%val{config}/kak-lsp.toml"
```

**Comment by krobelus** [comment]
Hmm I don't see anything wrong in there. Very strange.
If the problem goes away when disabling snippets, then it's likely something inside `lsp-snippets-insert-completion` (perhaps it doesn't handle the trailing `{` correctly?). Or `lsp_completion_inserted_ranges` is set to the wrong value?

There is also the option to run Kakoune [in a debugger](https://github.com/mawww/kakoune/wiki/Hacking#gdb) to see which completion is actually inserted, or if some completion is inserted twice etc., but that's unlikely to be the issue, so I'd only do that as last resort.

Since you mentioned that this bug existed for years, I think the `completionItem/resolve` requests are irrelevant.

**Comment by xnuk** [comment]
> then it's likely something inside `lsp-snippets-insert-completion`

Oh, setting `snippet_support = false` solved... the problem. (`snippet_support = true` can reproduce the problem, still)

**Comment by krobelus** [comment]
Huh the solution is very simple: you just need to upgrade kak-lsp.
Snippets had a long-standing bug that was finally fixed in kak-lsp 13.0.0 by https://github.com/kak-lsp/kak-lsp/pull/616
My bad, I misread the version - since `12.1.0-snapshot` is a snapshot version, my brain short-circuited and assumed that you're using latest master (there's no reason to use old snapshot versions).

**Comment by xnuk** [comment]
Oh upgrading kak-lsp solved the issue... for now. I thought `12.1.0-snapshot` is the latest version. I installed [`kak-lsp-git`](https://aur.archlinux.org/packages/kak-lsp-git) and forgot about it, so everyday's `pacman -Syu` didn't upgrade kak-lsp.... Thank you for digging my problem.
